google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/QueryRequest.java

```diff
@@ -102,6 +106,33 @@ public Builder setQuery(String query) {
     }
 
     /**
+     * Adds a QueryParameter to the list of query parameters. See
+     * {@link #setQueryParameters(Iterable)} for more details on the input requirements.
+     */
+    public Builder addQueryParameter(QueryParameter queryParameter) {
+      if (queryParameters == null) {
+        queryParameters = Lists.newArrayList();
+      }
+      queryParameters.add(checkNotNull(queryParameter));
+      return this;
+    }
+
+    /**
+     * Sets the query parameters for the query.
+     *
+     * The set of query parameters must either be all positional or all named parameters.
+     * Positional parameters are denoted in the query with a question mark (?), and named
```

Comment:
Indeed query must match and there is nothing we can do about it, however we could help with mixing them if that is not supported and also make it a cleaner API by not having an object with attributes that are not relevant and will always be null in that case.

I am fine with your suggestion if you plan to return Map<String, QueryParameterValue> for getNamedParameters and List<QueryParameterValue> for positional. I would prefer that over the current approach.